feat(cargo): support private crate

vmaerten opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes
For the Cargo manager, I've enabled by default [net.git-fetch-with-cli](https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli). 
I also enable hostRule for passing token to pull private crate (like it's done for Golang)


## Context
Closes #20059 
Cargo, in Renovate, could not pull private crate when updating artifacts. 
<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

I've updated the snapshot to see the new Cargo config, and I tested it locally pointing to a repo with a private crate.

It's my first PR, I hope I did not miss something.

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

secustor commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -8,21 +8,25 @@ import {
   }
 
   const execOptions: ExecOptions = {
+    extraEnv: { ...getGitEnvironmentVariables(['cargo`']) },
```
```suggestion
    extraEnv: { ...getGitEnvironmentVariables(['cargo']) },
```
?

Also test, that the env variables are applied.

secustor reviewed (changes requested):
Add an entry in in cargo/index.md that this is happening.

vmaerten commented:
Thanks for your review! 
I've updated the doc in cargo/readme.md, following the same convention as for golang (gomod/readme.md)
I've also added 3 tests to ensure that env variable are applied

HonkingGoose commented on docs/usage/rust.md:
```diff
@@ -31,9 +31,23 @@ Read the [Rust environment variables docs](https://doc.rust-lang.org/cargo/refer
-Third-party crate registries usually provide instructions to achieve this.
-
-Both of these are currently only possible when running Renovate self-hosted.
+This example shows how you can use a `hostRules` configuration to configure Renovate for use with a custom private Crate using Git to pull the modules when updating `Cargo.lock`.
```
This sentence does too much at once:

- Thing about `hostRules`
- Thing about custom private Crate using Git
- Pull modules when updating `Cargo.lock`

I can't keep all this in my head. So we should split the sentence into multiple parts.

HonkingGoose commented on docs/usage/rust.md:
```diff
@@ -31,9 +31,23 @@ Read the [Rust environment variables docs](https://doc.rust-lang.org/cargo/refer
-
-Both of these are currently only possible when running Renovate self-hosted.
+This example shows how you can use a `hostRules` configuration to configure Renovate for use with a custom private Crate using Git to pull the modules when updating `Cargo.lock`.
+All token `hostRules` with a `hostType` (e.g. `github`, `gitlab`, `bitbucket`, ... ) and host rules without a `hostType` are setup for authentication.
```
> are setup for authentication.

This is passive. I think we mean: "You as user can set authentication by doing x."

HonkingGoose commented on docs/usage/rust.md:
```diff
@@ -31,9 +31,23 @@ Read the [Rust environment variables docs](https://doc.rust-lang.org/cargo/refer
-Both of these are currently only possible when running Renovate self-hosted.
+This example shows how you can use a `hostRules` configuration to configure Renovate for use with a custom private Crate using Git to pull the modules when updating `Cargo.lock`.
+All token `hostRules` with a `hostType` (e.g. `github`, `gitlab`, `bitbucket`, ... ) and host rules without a `hostType` are setup for authentication.
+It's also possible to setup a hostRules which will be used only for Cargo authentication (e.g. `hostType: 'cargo'`).
```
```suggestion
You can configure a `hostRules` that's only for Cargo authentication (e.g. `hostType: 'cargo'`).
```

Let's be clear on _who_ does what, and simplify a bit.

HonkingGoose commented on lib/modules/manager/cargo/readme.md:
```diff
@@ -4,3 +4,20 @@ When using the default rangeStrategy=auto:
+
+Before running the `cargo` commands to update the `cargo.lock`, Renovate exports `git` [`insteadOf`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-urlltbasegtinsteadOf) directives in environment variables.
+
+The following logic is executed prior to "artifacts" updating:
```
```suggestion
Renovate uses this logic before it updates any "artifacts":
```

How about this?

vmaerten replied:
I totally agree with you.
I took this sentence in [lib/modules/manager/gomod/readme.md](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gomod/readme.md), It could be worth to align this between manager

HonkingGoose commented on lib/modules/manager/cargo/readme.md:
```diff
@@ -4,3 +4,20 @@ When using the default rangeStrategy=auto:
+The token from the `hostRules` entry matching `hostType=github` and `matchHost=api.github.com` is added as the default authentication for `github.com`.
+For those running against `github.com`, this token will be the default platform token.
+
+Next, all `hostRules` with both a token or username/password and `matchHost` will be fetched, except for any github.com one from above.
```
```suggestion
Next, all `hostRules` with both a token or username/password and `matchHost` will be fetched, except for any `github.com` one from above.
```

HonkingGoose reviewed (changes requested):
First pass review, needs more work after.

vmaerten commented:
@HonkingGoose I've applied your suggestions.
I also tried to rephase the self hosted part.

HonkingGoose reviewed (approved):
I'm happy with the docs, but I want a maintainer to review the left and right side of the Markdown diff. I suspect we're accidentally dropping some information.
